Spanner/src/RequestTrait.php

```diff
@@ -0,0 +1,249 @@
+<?php
+/**
+ * Copyright 2024 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner;
+
+use Google\Cloud\Spanner\Admin\Database\V1\Backup;
+use Google\Cloud\Spanner\Admin\Database\V1\CopyBackupMetadata;
+use Google\Cloud\Spanner\Admin\Database\V1\CreateBackupMetadata;
+use Google\Cloud\Spanner\Admin\Database\V1\CreateDatabaseMetadata;
+use Google\Cloud\Spanner\Admin\Database\V1\Database;
+use Google\Cloud\Spanner\Admin\Database\V1\OptimizeRestoredDatabaseMetadata;
+use Google\Cloud\Spanner\Admin\Database\V1\RestoreDatabaseMetadata;
+use Google\Cloud\Spanner\Admin\Database\V1\UpdateDatabaseDdlMetadata;
+use Google\Cloud\Spanner\Admin\Instance\V1\CreateInstanceConfigMetadata;
+use Google\Cloud\Spanner\Admin\Instance\V1\CreateInstanceMetadata;
+use Google\Cloud\Spanner\Admin\Instance\V1\Instance;
+use Google\Cloud\Spanner\Admin\Instance\V1\InstanceConfig;
+use Google\Cloud\Spanner\Admin\Instance\V1\UpdateInstanceConfigMetadata;
+use Google\Cloud\Spanner\Admin\Instance\V1\UpdateInstanceMetadata;
+use Google\Cloud\Spanner\Session\SessionPoolInterface;
+use Google\Protobuf\GPBEmpty;
+
+/**
+ * Shared functionality for Spanner requests.
+ *
+ * @internal
+ */
+trait RequestTrait
+{
+    private $larHeader = 'x-goog-spanner-route-to-leader';
+    private $resourcePrefixHeader = 'google-cloud-resource-prefix';
+
+    /**
+     * Create request and send the request.
+     */
+    private function createAndSendRequest(
+        string $clientClass,
+        string $method,
+        array $data,
+        array $optionalArgs,
+        string $requestClass,
+        string $resourcePrefixHeader = '',
+        bool $routeToLeader = false
+    ) {
+        if ($resourcePrefixHeader) {
+            $optionalArgs = $this->addResourcePrefixHeader(
+                $optionalArgs,
+                $resourcePrefixHeader
+            );
+        }
+        if ($routeToLeader) {
+            $optionalArgs = $this->addLarHeader($optionalArgs, $routeToLeader);
+        }
+
+        $request = $this->serializer->decodeMessage(new $requestClass(), $data);
+
+        return $this->requestHandler->sendRequest(
+            $clientClass,
+            $method,
+            $request,
+            $optionalArgs
+        );
+    }
+
+    /**
+     * Retrun the LRO mappers.
+     *
+     * @return array
+     */
+    private function getLROResponseMappers()
```

Comment:
Removing duplicate data in the handwritten layer that's already being generated and handled by GAPIC is the whole reason we are making the V2 changes, so having all this metadata defined in an array like this is not acceptable.

We can consume the generated data in the GAPIC layer, or use the GAPIC clients directly.